keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,425 @@
+**How Resource classes can solve this:**  Property/attribute which forms the grouping can be advertised in the device attributes and then a resource can be created to form a grouped super-resource based on that property.<br/>
+**Can this be solved without resource classes:** No
+
+- I want to have quota control on the devices at the granularity of device properties. For example, I want to have a separate quota for ECC enabled GPUs. I want a specific user to not let use more than ‘N’ number of ECC enabled GPUs overall at namespace level.<br/>
+**Motivation:**  This will make it possible to charge user per specialized hw consumption. Since special HW is costly, as an Operator I want to have this capability.<br/>
+**How Resource classes can solve this:** Quota will be supported on resource class objects and by allowing resource request in user pods via resource class, charging policy can be linked with resource consumption.<br/>
+**Can this be solved without resource classes:** No
+
+- In my cluster, I have many different classes (different capabilities) of a device type (ex: NICs). End user’s expectations are met as long as device has a very small subset of these capabilities. I want a mechanism where end user can request devices which satisfies their minimum expectation.
+Few nodes are connected to data network over 40 Gig NICs and others are connected over normal 1 Gig NICs. I want end user pods to be able to request
+data network connectivity with high network performance while
+in default case, data network connectivity is offered via normal 1 Gbps NICs.<br/>
+**Motivation:** If some workloads demand higher network bandwidth, it should be possible to run these workloads on selected nodes.<br/>
+**Can this be solved without resource classes:** Taints and tolerations can help in steering pods but the problem in that there is no way today to have access control over use of tolerations and therefore if multiple users are there, it is not possible to have control on allowed tolerations.<br/>
+**How Resource classes can solve this:** I can define a ResourceClass for the high-performance NIC with minimum bandwidth requirements, and makes sure only users with proper quota can use such resources.
+
+- I want to be able to utilize different 'types' of a HW resource (not necessarily from the same vendor) while not losing workload portability when moving from one cluster/cloud to another. There can be one type of Nvidia GPUs on one cluster and another type of Nvidia GPUs on another cluster. This is example of different ‘types’ of a HW resource(GPU). I want to offer GPUs to be consumed under a same portable name, as long as their capabilities are almost same. If pods are consuming these GPUs with a generic resource class name, workload can be migrated from one cluster to another transparently.<br/>
+
+**Quoting Henry Saputra (From Ebay) for the record:** <br/>
```

Comment:
The issue isn't workload portability but the lack of standard way to expose resource attributes leading to having non standard way of declaring and exposing resources to the user.

Reply:
Providing a standard way to expose HW resource attributes is part of this design. As mentioned, a main motivation of this proposal is to provide a building block for vendors to export their HW with special properties on k8s through extension plugins. However, as we are providing more extension points, we do need to be careful that workload experience will not become fragmented across different environments. That is why another big part of the design is on defining an abstraction layer that resource providers and cluster admins can use to hide the underlying HW details. I know from vendor point of view, extensibility is more important than portability, but workload portability is important to customers. I think it has been a major goal of Kubernetes that we allow flexible extensions without sacrificing portability.

Also I hope our discussions are not just limited to GPU. For other non-primary resources like high-performance NIC, it is more likely that devices from different vendors may offer similar workload experience.

Reply:
> Also I hope our discussions are not just limited to GPU. For other non-primary resources like high-performance NIC, it is more likely that devices from different vendors may offer similar workload experience.

Please provide these examples then.

> Defining an abstraction layer that resource providers and cluster admins can use to hide the underlying HW details.

We already went over this issue at the F2F. It doesn't make sense to try to abstract / hide HW details. Your code is usually already compiled for that specific hardware.
If it is a real use case then it would be great to have real world example. I have yet to see a real world use case of a user requesting a "fast NIC" and have that mean a different hardware on two different cluster.

In any case, my point here is that the issue raised by Henry here is not that he needs to abstract HW details. He needs Kubernetes to expose a standard way of declaring and exposing resources and their attributes.